Dockerfile

```diff
@@ -1,34 +1,60 @@
-FROM debian:stable-slim as fetcher
+# Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
 
-# TODO copy just the 'distribution' directory, not all source code
+FROM ubuntu:bionic as build
+
+RUN apt-get update && apt-get install -y openjdk-8-jdk maven python3 python3-venv
+
+# Create a first layer to cache the "Maven World" in the local repository.
+# Incremental docker builds will always resume after that, unless you update the pom
+WORKDIR /mvn
+COPY pom.xml /mvn/
+COPY opengrok-indexer/pom.xml /mvn/opengrok-indexer/
+COPY opengrok-web/pom.xml /mvn/opengrok-web/
+COPY plugins/pom.xml /mvn/plugins/
+COPY suggester/pom.xml /mvn/suggester/
+
+# distribution and opengrok-tools do not have dependencies to cache
+RUN sed -i 's:<module>distribution</module>::g' /mvn/pom.xml
+RUN sed -i 's:<module>opengrok-tools</module>::g' /mvn/pom.xml
+
+RUN mkdir -p /mvn/opengrok-indexer/target/jflex-sources
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp
+COPY opengrok-indexer/jflex-tt.txt /mvn/opengrok-indexer/jflex-tt.txt
+COPY opengrok-indexer/jflex-tt-end.txt /mvn/opengrok-indexer/jflex-tt-end.txt
+COPY opengrok-indexer/jflex-code.txt /mvn/opengrok-indexer/jflex-code.txt
+COPY opengrok-indexer/jflex-code-end.txt /mvn/opengrok-indexer/jflex-code-end.txt
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp/WEB-INF/ && touch /mvn/opengrok-web/src/main/webapp/WEB-INF/web.xml
```

Comment:
I think this section here actually makes it necessary to run Docker builds for every push/PR as it relies on internal structure of the project.

Reply:
Yes, you are right :D I always forget how overcomplicated the build is :D

Reply:
more than before I mean.